In [1]:
 %config Completer.use_jedi = False                                       
import warnings                                                             
warnings.filterwarnings(action="ignore")
import numpy as np                                                          
import pandas as pd 

***
평균 제곱근 오차(RMSE, Root Mean Squared Error)
***
임의의 선을 그리고 난 후 이 선이 얼마나 잘 그려졌는지 평가해서 조금씩 수정해나가기 위해 주어진 선의 오차를 평가하는 알고리즘이 필요한데,  
그 중 가장 많이 사용되는 알고리즘이 "평균 제곱근 오차"이다.

In [2]:
# 앞의 예제 파일 '04_기울기_y절편'에서 xData, yData 데이터 값이 1건씩 저장된 리스트
#[xData(공부시간), yData(공부시간에 따른 시험점수)]
data = [ [2, 81], [4, 93], [6, 91], [8, 97] ]
# 기울기 a와 y절편 b를 임의로 정한다. -> [기울기, y절편] -> [2.3, 79]가 최적 기울기와 y절편
ab = [2.3, 79]

for i in data:
    print('공부시간 : {}시간, 공부시간에 따른 시험점수 : {}점'.format( i[0], i[1] ))


공부시간 : 2시간, 공부시간에 따른 시험점수 : 81점
공부시간 : 4시간, 공부시간에 따른 시험점수 : 93점
공부시간 : 6시간, 공부시간에 따른 시험점수 : 91점
공부시간 : 8시간, 공부시간에 따른 시험점수 : 97점


In [3]:
#공부시간과 공부시간에 따른 시험점수 데이터 값이 1건씩 저장된 리스트에서 공부시간과 공부시간에 따른 시험점수를 뽑아내
#별도의 리스트에 저장한다.
xData = [i[0] for i in data]
yData = [i[1] for i in data]

for i in range(len(xData)):
    print('공부시간 : {}시간, 공부시간에 따른 시험점수 : {}점'.format( xData[i], yData[i] ))

공부시간 : 2시간, 공부시간에 따른 시험점수 : 81점
공부시간 : 4시간, 공부시간에 따른 시험점수 : 93점
공부시간 : 6시간, 공부시간에 따른 시험점수 : 91점
공부시간 : 8시간, 공부시간에 따른 시험점수 : 97점


In [4]:
# Y = ax + b 수식에 기울기, y절편 값을 대입해서 결과 (예측성적)를 계산하는 함수
def predict(x):
    return (ab[0] * x + ab[1])

In [5]:
#임의로 정한 기울기와 y절편으로 예측성적을 계산한다.
predict_result = [] # 예측성적을 저장할 빈 리스트
#모든 공부시간 (xData)값을 1번씩 대입해서 예측 성적 리스트(predict_result)에 추가한다.
for i in range(len(xData)):
    predict_result.append( predict(xData[i]) )
    print('공부시간 : {}시간, 공부시간에 따른 시험점수 : {}점, 예측성적 : {}점'.format( xData[i], yData[i], predict_result[i] ))
# ------ for i

공부시간 : 2시간, 공부시간에 따른 시험점수 : 81점, 예측성적 : 83.6점
공부시간 : 4시간, 공부시간에 따른 시험점수 : 93점, 예측성적 : 88.2점
공부시간 : 6시간, 공부시간에 따른 시험점수 : 91점, 예측성적 : 92.8점
공부시간 : 8시간, 공부시간에 따른 시험점수 : 97점, 예측성적 : 97.4점


In [6]:
#평균 제곱근 오차를 계산하는 함수 -> rmse(예측성적, 실제성적) -> 인수로 넘겨받는 데이터는 numpy 배열 타입의 데이터이어야 한다.
def rmse(predict_result, yData):
    # 파이썬의 리스트는 브로드캐스팅 연산이 불가능하므로 numpy의 배열로 바꿔서 연산한다.
    predict_result = np.array(predict_result)
    yData = np.array(yData)
    return np.sqrt(((predict_result - yData) ** 2).mean())

In [8]:
#평균 제곱근 오차를 계산해서 출력한다.
rmse_result = rmse(predict_result, yData)
print('RMSE -> {}'.format( rmse_result ))

RMSE -> 2.880972058177584


***
기울기 -> 3, y절편 -> 76일 경우 RMSE는 3.3166247903554가 계산되는 것을 확인할 수 있다.  
기울기 -> 2.1, y절편 -> 82일 경우 RMSE는 3.535533905932736 계산되는 것을 확인할 수 있다.  
이 오차를 줄이면서 새로운 선을 긋는 작업을 해야한다.  
이를 위해서는 기울기와 y절편을 적절히 조절하면서 오차의 변화를 살펴보고, 오차가 최소가 되는  
기울기와 y절편을 찾아야한다.  
***
앞의 예제 파일 '04_기울기_y절편'에서 계산된 기울기 -> 2.3, y절편 -> 79.0일 경우 RMSE는 2.880972058177584가 계산된다.  

In [9]:
#[2, 4, 6, 8] - [81, 93, 91, 97] #에러!!!
np.array([2, 4, 6, 8]) - np.array([81, 93, 91, 97]) #numpy 배열은 가능!!!!

array([-79, -89, -85, -89])